napari/components/_layer_slicer.py

```diff
@@ -0,0 +1,138 @@
+
+from napari.components import Dims
+from napari.layers import Layer
+from napari.utils.events.event import EmitterGroup, Event
+
+logger = logging.getLogger("napari.components._layer_slicer")
+
+
+class _LayerSlicer:
+    """
+    High level class to control the creation of a slice (via a slice request),
+    submit it (synchronously or asynchronously) to a thread pool, and emit the
+    results when complete.
+
+    Events
+    ------
+    ready
+        emitted after slicing is done with a dict value that maps from layer to slice response
+    """
+
+    def __init__(self):
+        """
+        Attributes
+        ----------
+        _executor : concurrent.futures.ThreadPoolExecutor
+            manager for the slicing threading
+        _layers_to_task : dict
+            task storage for cancellation logic
+        _lock_layers_to_task : threading.RLock
+            lock to guard against changes to `_layers_to_task` when finding,
+            adding, or removing tasks.
+        """
+        self.events = EmitterGroup(source=self, ready=Event)
+        self._executor: Executor = ThreadPoolExecutor(max_workers=1)
+        self._layers_to_task: Dict[Tuple[Layer], Future] = {}
+        self._lock_layers_to_task = RLock()
+
+    def slice_layers_async(
+        self, layers: Iterable[Layer], dims: Dims
+    ) -> Future[dict]:
+        """This should only be called from the main thread.
+
+        Creates a new task and adds it to the _layers_to_task dict. Cancels
+        all tasks currently pending for that layer tuple.
+
+        Submitting multiple layers at one generates multiple requests (stored in a dict),
+        but only ONE task.
+
+        A slice submitted with multiple layers will only cancel if the same tuple of layers
+        is resubmitted. If a second slice is requested with only one layer from the first
+        slice, it will not be cancelled.
+        """
+        # Cancel any tasks that are slicing a subset of the layers
+        # being sliced now. This allows us to slice arbitrary sets of
+        # layers with some sensible and not too complex cancellation
+        # policy.
+        if existing_task := self._find_existing_task(layers):
+            logger.debug('Cancelling task for %s', layers)
+            existing_task.cancel()
+
+        # Not all layer types will initially be asynchronously sliceable.
+        # The following logic gives us a way to handle those in the short
+        # term as we develop, and also in the long term if there are cases
+        # when we want to perform sync slicing anyway.
+        requests = {}
+        for layer in layers:
+            if layer._is_async():
+                requests[layer] = layer._make_slice_request(dims)
+            else:
+                layer._slice_dims(dims.point, dims.ndisplay, dims.order)
+        # create task for slicing of each request/layer
+        task = self._executor.submit(self._slice_layers, requests)
```

Comment:
This was [previously discussed in the NAP and is high priority future work](https://napari.org/dev/naps/4-async-slicing.html#render-each-slice-as-soon-as-it-is-ready), but we made it out of scope for this initial work because it required some more thought on the design/front-end behavior (e.g. what do you show when you have one layer slicer, but are waiting for another?).

As a reminder, the main goal of this work is to free up the main thread to not block so that napari continues to respond to user input. Running multiple slicing requests parallel should speed up slicing of multiple layer (in some cases), but it's more like an optimization.

Reply:
I will wait for all slices to be done before performing any action.

Reply:
Sorry, I don't understand this comment.

Do you mean that's your personal preference? If so, we plan to support ways to force sync slicing both permanently (e.g. through a user setting) and temporarily (e.g. with a context manager).

Or are you suggesting a different implementation here?

Reply:
Agreed that's the safe, desired behavior most of the time. That's effectively what we're doing in this PR (at least for the async slice tasks), but without the potential optimization of parallel read requests.

I think that this PR offers a potential improvement (when it's actually integrated) over `main` (where we trigger things like rendering in the order of the layers all on the main thread).

Adding support for parallel read requests (e.g. one thread/task per layer) is doable, but my instinct is to add that in a follow-up PR since it complicates scheduling and cancellation logic. What do you think?